[docu-2457] Router Reference, Router Route creation (#4251)
Kong/docs.konghq.com: src/gateway/understanding-kong/how-to/router-atc.md

---
title: Configure Routes with Expressions
content-type: how-to
---

> With the release of version 3.0, {{site.base_gateway}} now ships with a new router. The new router can describe routes using a domain-specific language called Expressions. Expressions can describe routes or paths as patterns using regular expressions. This how-to guide will walk through switching to the new router, and configuring routes with the new expressive domain specific language, expressions. For a list of all available operators and configurable fields please review the [reference documentation](/gateway/latest/reference/router-operators).

Reviewer: ```suggestion
With the release of version 3.0, {{site.base_gateway}} now ships with a new router. The new router can describe routes using a domain-specific language called Expressions. Expressions can describe routes or paths as patterns using regular expressions. 

This how-to guide will walk through switching to the new router and configuring routes with the new expressive domain specific language, expressions. For a list of all available operators and configurable fields please review the [reference documentation](/gateway/latest/reference/router-operators).
```
Same questions here about capitalizing "Expressions" and 3.0 vs. 3.0.x

## Prerequisite

> * Edit [kong.conf](/gateway/latest/kong-production/kong-conf) to contain the line `router_flavor = atc` and restart {{site.base_gateway}}.

Reviewer: ```suggestion
Edit [kong.conf](/gateway/latest/kong-production/kong-conf) to contain the line `router_flavor = atc` and restart {{site.base_gateway}}.
```

## Create routes with Expressions

> To create a new router object using expressions send a `POST` request to the [services endpoint](/gateway/latest/admin-api/#update-route) like this: 

Reviewer: ```suggestion
To create a new router object using expressions, send a `POST` request to the [services endpoint](/gateway/latest/admin-api/#update-route) like this: 
```

```sh
curl --request POST \
  --url http://localhost:8001/services/example-service/routes \
  --form  atc='http.path == "/mock"
```

> In this example you associated a new route object with the path `/mock` to the existing service `example-service`. The expressions DSL allows you to create complex router objects using operators and fields. 

Reviewer: ```suggestion
In this example, you associated a new route object with the path `/mock` to the existing service `example-service`. 

The expressions DSL allows you to create complex router objects using operators and fields. 
```
I think that last sentence should be on a new line? I'm also not sure what "DSL" stands for.

```sh
curl --request POST \
  --url http://localhost:8001/services/example-service/routes \
  --header 'Content-Type: multipart/form-data' \
  --form 'atc=(http.path == "/mock" || net.protocol == "https")'
```
In this example the || operator created an expression that set variables for the following fields: 

```
"protocols": ["http", "https"]
AND
"paths": ["/mock"]
```

> You can use non-expressions related attributes within the same `POST` request:

Reviewer: ```suggestion
You can use attributes that are unrelated to expressions within the same `POST` request:
```
Just reworked it a bit because I think it would be "non-expressions-related" technically.

```sh
curl --request POST \
  --url http://localhost:8001/services/example-service/routes \
  --header 'Content-Type: multipart/form-data' \
  --form 'atc=(http.path == "/mock" || net.protocol == "https") \
  --form name=mocking
```

This would define a router object with the following attributes:

```
"atc": "(http.path == \"\/mock\" || net.protocol == \"https\"),
"name": "mocking",

```

> The list of available attributes is available in the [reference documentation](/gateway/latest/admin-api/#request-body). 

Reviewer: ```suggestion
You can view the list of available attributes in the [reference documentation](/gateway/latest/admin-api/#request-body). 
```
Just rewrote it to avoid multiple repetitions of "available"

### Create complex routes with Expressions

You can describe complex route objects using operators within a `POST` request. 

```sh

curl --request POST \
  --url http://localhost:8001/services/example-service/routes \
  --header 'Content-Type: multipart/form-data' \
  --form name=complex_object \
  --form 'atc=(net.protocol == "http" || net.protocol == "https") &&
         (http.method == "GET" || http.method == "POST") &&
         (http.host == "example.com" || http.host == "example.test") &&
         (http.path ^= "/mock" || http.path ^= "/mocking") &&
         http.headers.x_another_header == "example_header" && (http.headers.x_my_header == "example" || http.headers.x_my_header == "example2")'
```

> This request returns a `200` with the following values: 

Reviewer: ```suggestion
This request returns a `200` status code with the following values: 
```

```sh
"protocols": ["http", "https"],
"methods": ["GET", "POST"],
"hosts": ["example.com", "example.test"],
"paths": ["/mock", "/mocking"],
"headers": {"x-another-header":["example_header"], "x-my-header":["example", "example2"]},
```

> For a list of all available operators refer to the [reference documentation](/gateway/latest/reference/router-operators.

Reviewer: ```suggestion
For a list of all available operators, see the [reference documentation](/gateway/latest/reference/router-operators/).
```

## More information

> [Expressions repository](https://github.com/Kong/atc-router#table-of-contents)
> [Expressions Reference](/gateway/latest/reference/router-operators)

Reviewer: ```suggestion
* [Expressions repository](https://github.com/Kong/atc-router#table-of-contents)
* [Expressions Reference](/gateway/latest/reference/router-operators)
```
